fix(linter/plugins): resolve JS plugins only with conditions Node.js supports

sapphi-red opened the pull request:
Fixes the `require` related error described in https://github.com/oxc-project/oxc/discussions/14862#discussioncomment-14861678.
This was happening because Node.js does not use `module` condition but the plugin resolver used that. I aligned the conditions to what Node.js uses (I didn't include `node-addon` because it's not popular and requires detecting `--no-addons` to do it properly).
https://nodejs.org/docs/v24.11.0/api/packages.html#conditional-exports

I'm not sure why, but the test snapshot doesn't seem stable on my machine.

cc @TheAlexLichter 

refs #14541

sapphi-red commented:
* **#15248** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/15248?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/15248?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=15248).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

camc314 commented:
> I'm not sure why, but the test snapshot doesn't seem stable on my machine.

possibly `pnpm run build-dev` vs `pnpm run build-test`, build test enables a special feature flag for more consistent diagnostics. i'll take a look

camc314 commented:
### Merge activity

* **Nov 4, 3:52 PM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

sapphi-red committed:
fix(linter/plugins): resolve JS plugins only with conditions Node.js supports (#15248)

Fixes the `require` related error described in https://github.com/oxc-project/oxc/discussions/14862#discussioncomment-14861678.
This was happening because Node.js does not use `module` condition but the plugin resolver used that. I aligned the conditions to what Node.js uses (I didn't include `node-addon` because it's not popular and requires detecting `--no-addons` to do it properly).
https://nodejs.org/docs/v24.11.0/api/packages.html#conditional-exports

I'm not sure why, but the test snapshot doesn't seem stable on my machine.

cc @TheAlexLichter

refs #14541
